### PR TITLE
adds padding when num poly wires is different

### PR DESCRIPTION
**Pull request by ziofil**
### **User description**
We allow making linear superpositions of components with different number of polynomial wires by padding A and b with zeros, and adding empty dimensions to the c array.

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Enhanced the addition operation in `ansatze.py` to pad matrices `A` and `b` when the number of polynomial wires differs, ensuring compatibility.
- Introduced a check to raise a `ValueError` if polynomial shapes are inconsistent.
- Expanded dimensions of arrays to handle different shapes during addition.
- Added tests to verify correct padding of `A` and `b` and to ensure a `ValueError` is raised for inconsistent polynomial shapes.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>ansatze.py</strong><dd><code>Add padding for polynomial wires in ansatze addition</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

mrmustard/physics/ansatze.py

<li>Added padding for <code>A</code> and <code>b</code> matrices when polynomial wires differ.<br> <li> Implemented checks for consistent polynomial shapes.<br> <li> Expanded dimensions of arrays to accommodate different shapes.<br>

</details>

  </td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/499/files#diff-85405583f9e1c800d7cbad37cca5f7da9c0181a18dd40db2a207dcd134928fe0">+18/-1</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_ansatz.py</strong><dd><code>Add tests for polynomial wire padding and shape consistency</code></dd></summary>
<hr>

tests/test_physics/test_ansatz.py

<li>Added test for padding <code>A</code> and <code>b</code> with different polynomial wires.<br> <li> Added test for inconsistent polynomial shapes raising a ValueError.<br>

</details>

  </td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/499/files#diff-8ae3f2522d16e5dd99914d95d64920f61fb075eb99cec87ee2a122872331da59">+31/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

File: mrmustard/physics/ansatze.py [hunk 1]
```diff
@@ -417,8 +417,25 @@ def __add__(self, other: PolyExpBase) -> PolyExpBase:
...
-        shape of the combined object will be (2,5,4,5).
+        shape of the combined object will be (2,5,4,5). It also pads A and b, to account for an eventual
+        different number of polynomial wires.
         """
+        (_, n, _) = self.mat.shape
+        (_, m, _) = other.mat.shape
+        self_num_poly, _ = self.polynomial_shape
+        other_num_poly, _ = other.polynomial_shape
+        if self_num_poly - other_num_poly != n - m:
+            raise ValueError("Inconsistent polynomial shapes.")
```
Comment (apchytr) [inline mrmustard/physics/ansatze.py:428]: Can you include printing the polynomial shapes?

File: mrmustard/physics/ansatze.py [hunk 2]
```diff
@@ -417,8 +417,25 @@ def __add__(self, other: PolyExpBase) -> PolyExpBase:
         Adds two ansatze together. This means concatenating them in the batch dimension.
         In the case where c is a polynomial of different shapes it will add padding zeros to make
         the shapes fit. Example: If the shape of c1 is (1,3,4,5) and the shape of c2 is (1,5,4,3) then the
-        shape of the combined object will be (2,5,4,5).
+        shape of the combined object will be (2,5,4,5). It also pads A and b, to account for an eventual
+        different number of polynomial wires.
         """
+        (_, n, _) = self.mat.shape
+        (_, m, _) = other.mat.shape
+        self_num_poly, _ = self.polynomial_shape
```
Comment (apchytr) [inline mrmustard/physics/ansatze.py:425]:
```suggestion
        self_num_poly, a0s = self.polynomial_shape
```
then you can remove `442` below. Ditto for `other_num_poly`.

File: mrmustard/physics/ansatze.py [hunk 3]
```diff
@@ -417,8 +417,25 @@ def __add__(self, other: PolyExpBase) -> PolyExpBase:
...
+        other_num_poly, _ = other.polynomial_shape
+        if self_num_poly - other_num_poly != n - m:
+            raise ValueError("Inconsistent polynomial shapes.")
+        if n < m:
+            self.mat = math.pad(self.mat, ((0, 0), (0, m - n), (0, m - n)))
```
Comment (apchytr) [inline mrmustard/physics/ansatze.py:430]: I believe these are mutating `self` and `other` (below). Is this what we want? I would be a bit worried about having unintended effects in this case. Maybe we should use similar logic for how we handle `c` instead i.e. a0_new, a1_new

Comment (apchytr) [reply mrmustard/physics/ansatze.py:430]: Also, are we able to combine this logic with the `c` padding logic below in any way? That should clean some things up

Comment (ziofil) [reply mrmustard/physics/ansatze.py:430]: Hm, yes maybe. Thanks for the suggestion, I'll see what I can do. 👍

File: mrmustard/physics/ansatze.py [hunk 4]
```diff
@@ -417,38 +417,47 @@ def __add__(self, other: PolyExpBase) -> PolyExpBase:
         Adds two ansatze together. This means concatenating them in the batch dimension.
         In the case where c is a polynomial of different shapes it will add padding zeros to make
         the shapes fit. Example: If the shape of c1 is (1,3,4,5) and the shape of c2 is (1,5,4,3) then the
-        shape of the combined object will be (2,5,4,5).
+        shape of the combined object will be (2,5,4,5). It also pads A and b, to account for an eventual
+        different number of polynomial wires.
         """
-        combined_matrices = math.concat([self.mat, other.mat], axis=0)
-        combined_vectors = math.concat([self.vec, other.vec], axis=0)
+        (_, n1, _) = self.mat.shape
```
Comment (apchytr) [inline mrmustard/physics/ansatze.py:423]: Might be cleaner to just do `n1 = self.mat.shape[1]`

**Review by apchytr — APPROVED** [review]
Looks great!

**Pull request merged**